fix: disallow undefined where clause

anikdhabal opened the pull request:
## What does this PR do?

## Summary by mrge
Disallowed undefined, empty object, and empty array values in Prisma where clauses to prevent invalid queries.

- **Bug Fixes**
  - Added validation to throw errors if a where clause is undefined, empty, or contains undefined fields.
  - Updated extension to cover findMany, updateMany, and deleteMany queries.

<!-- End of auto-generated description by mrge. -->

anikdhabal committed:
fix: disallow undefined where clause

anikdhabal committed:
Merge branch 'main' into where-caluse

anikdhabal committed:
update

anikdhabal commented on packages/prisma/extensions/disallow-undefined-where-clause.ts:
```diff
@@ -24,16 +39,20 @@ export const checkUndefinedInValue = (where: any) => {
     query: {
       $allModels: {
         async deleteMany({ args, query }) {
-          checkUndefinedInValue(args.where);
+          validateWhereClause(args.where);
           return query(args);
         },
         async updateMany({ args, query }) {
-          checkUndefinedInValue(args.where);
+          validateWhereClause(args.where);
+          return query(args);
+        },
+        async findMany({ args, query }) {
+          validateWhereClause(args.where);
           return query(args);
```
Currently, we have this guard for these operations. Do we need to add more operations here?

anikdhabal committed:
Update disallow-undefined-where-clause.ts

anikdhabal commented on packages/prisma/extensions/disallow-undefined-where-clause.ts:
```diff
@@ -1,6 +1,21 @@
+  }
+
+  // Check if where is an empty object
+  if (typeof where === "object" && !Array.isArray(where) && Object.keys(where).length === 0) {
```
Fixed

anikdhabal committed:
fix test

anikdhabal committed:
Merge branch 'where-caluse' of https://github.com/calcom/cal.com into where-caluse

anikdhabal committed:
another one

anikdhabal commented on packages/prisma/extensions/disallow-undefined-where-clause.test.ts:
```diff
@@ -0,0 +1,91 @@
+    );
+  });
+
+  it("validateWhereClause should throw exception when the where object is undefined", async () => {
+    const where = undefined;
+
+    expect(() => validateWhereClause(where)).toThrowError('The "where" clause cannot be undefined.');
+  });
+
+  it("validateWhereClause should throw exception when the where object is {}", async () => {
+    const where = {};
+
+    expect(() => validateWhereClause(where)).toThrowError('The "where" clause cannot be an empty object {}.');
+  });
+
+  it("validateWhereClause should throw exception when the where object is []", async () => {
+    const where = [];
+
+    expect(() => validateWhereClause(where)).toThrowError('The "where" clause cannot be an empty array [].');
+  });
+
+  it("validateWhereClause should throw exception when the 'in' field of where object is []", async () => {
+    const where = {
+      id: {
+        in: [],
+      },
+    };
+
+    expect(() => validateWhereClause(where)).toThrowError(
+      'The "in" value for the field "id" cannot be an empty array [].'
+    );
+  });
+});
```
Added new test

anikdhabal committed:
Merge branch 'main' into where-caluse

Udit-takkar reviewed (changes requested):
All E2E tests are failing

anikdhabal committed:
Merge branch 'main' into where-caluse

anikdhabal committed:
fix test

anikdhabal committed:
Merge branch 'main' into where-caluse

anikdhabal committed:
Merge branch 'main' into where-caluse

anikdhabal committed:
update

Udit-takkar committed:
Merge branch 'main' into where-caluse

devin-ai-integration[bot] committed:
fix: resolve 29 e2e test failures caused by Prisma extension

- Add proper where clauses to all Prisma operations in e2e test fixtures
- Fix mock Prisma to use empty object instead of undefined where clause
- Update test helpers to include where clauses in findMany operations
- Maintain Prisma extension security functionality while fixing test compatibility

Co-Authored-By: anik@cal.com <adhabal2002@gmail.com>

keithwillcode commented on apps/web/playwright/fixtures/features.ts:
```diff
@@ -13,7 +13,9 @@ export const createFeatureFixture = (page: Page) => {
     init: async () => {
-      const features = await prisma.feature.findMany();
+      const features = await prisma.feature.findMany({
+        where: {},
```
Not a fan of having to do this

anikdhabal committed:
Merge branch 'main' into where-caluse

anikdhabal committed:
update

anikdhabal committed:
update

anikdhabal committed:
revert

volnei commented:
Closing it in favor of `strictUndefinedChecks`. Requires prisma upgrade.
